src/Autofac/Core/Resolving/ResolveOperationBase.cs

```diff
@@ -124,7 +124,17 @@ protected ResolveOperationBase(ISharingLifetimeScope mostNestedLifetimeScope, IR
         /// Don't want this exposed to the outside world, but we do want it available in the <see cref="CircularDependencyDetectorMiddleware"/>,
         /// hence it's internal.
         /// </remarks>
-        internal Stack<ResolveRequestContextBase> RequestStack { get; } = new Stack<ResolveRequestContextBase>();
+        internal SegmentedStack<ResolveRequestContextBase> RequestStack { get; } = new SegmentedStack<ResolveRequestContextBase>();
+
+        /// <summary>
+        /// Enter a new dependency chain block where subsequent requests inside the operation are allowed to repeat
+        /// registrations from before the block.
+        /// </summary>
+        /// <returns>A disposable that should be disposed to exit the block.</returns>
+        public IDisposable EnterNewDependencyDetectionBlock()
+        {
+            return RequestStack.EnterSegment();
```

Comment:
I'm noticing `RequestDepth` is getting tracked separately and used in, say, `GetOrCreateInstance`. Does this new stack mechanism make that separate tracking no longer required?

Reply:
Possibly yes, although I would need to add a `GlobalCount` to the `SegmentStack`, because `Count` only gives the size of the current segment.

Might be worth it to avoid the extra increment/decrement instructions.

Reply:
Ah, I remember now why it's tracked separately; property injection that permits circular dependencies will always have an empty request stack, because it runs in Complete, but still needs the depth detection counted separately. Otherwise we get a StackOverflowException.